internal/daemon/api_logs.go

```diff
@@ -0,0 +1,262 @@
+		for i, info := range infos {
+			services[i] = info.Name
+		}
+	}
+
+	// Get log iterators by service (and close them when we're done).
+	itsByName, err := r.svcMgr.ServiceLogs(services, numLogs)
+	if err != nil {
+		response := statusInternalError("cannot fetch log iterators: %v", err)
+		response.ServeHTTP(w, req)
+		return
+	}
+	defer func() {
+		for _, it := range itsByName {
+			_ = it.Close()
+		}
+	}()
+
+	// Make sorted list of service names we actually have iterators for.
+	services = services[:0]
+	for name := range itsByName {
+		services = append(services, name)
+	}
+	sort.Strings(services)
+
+	// Create an iterator and log parser for each service.
+	iterators := make([]servicelog.Iterator, len(services))
+	parsers := make([]*servicelog.Parser, len(services))
+	for i, name := range services {
+		iterators[i] = itsByName[name]
+		parsers[i] = servicelog.NewParser(iterators[i], logReaderSize)
+	}
+
+	// Output format is JSON Lines, which doesn't have an official mime type,
+	// but "application/x-ndjson" is what most people seem to use:
+	// https://github.com/wardi/jsonlines/issues/9
+	w.Header().Set("Content-Type", "application/x-ndjson")
+	encoder := json.NewEncoder(w)
+	encoder.SetEscapeHTML(false)
+	requestDone := req.Context().Done()
+	requestStarted := time.Now().UTC()
+
+	// If we'll be following real-time, make a channel and register it with
+	// each of the iterators to be notified when new data comes in.
+	var notification chan bool
+	if follow {
+		notification = make(chan bool, 1)
+		for _, it := range iterators {
+			it.Notify(notification)
+		}
+	}
+
+	// Use a buffered channel as a FIFO for keeping the latest numLogs logs if
+	// request "n" is set (the default).
+	var fifo chan servicelog.Entry
+	if numLogs > 0 {
+		fifo = make(chan servicelog.Entry, numLogs)
+	}
+
+	// Slice of next entries for each service
+	nexts := make([]servicelog.Entry, len(services))
+
+	// Changes to true once we start following
+	following := false
+
+	// Main loop: output earliest log per iteration. Stop when no more logs,
+	// or when request is cancelled if in follow mode.
+	for {
+		// Try to fetch next log from each service (parser/iterator combo).
+		for i, parser := range parsers {
+			if !nexts[i].Time.IsZero() {
+				continue
+			}
+			if parser.Next() {
+				nexts[i] = parser.Entry()
+			} else if parser.Err() != nil {
+				log.Printf("error parsing logs: %v", parser.Err())
+				return
+			} else if iterators[i].Next(nil) {
+				parser.Reset()
+			}
+		}
+
+		// Find the log with the next earliest timestamp.
+		earliest := -1
+		for i, next := range nexts {
+			if next.Time.IsZero() {
+				continue
+			}
+			if earliest < 0 || next.Time.Before(nexts[earliest].Time) {
+				earliest = i
+			}
+		}
+
+		// No more logs: flush and exit, or wait for another if in follow mode.
+		if earliest < 0 {
+			// If they requested "n" logs, flush logs in the FIFO first.
+			if numLogs > 0 && len(fifo) > 0 {
```

Comment:
This new loop is already looking pretty nice in the sense we have the logic almost entirely agnostic to what form of iteration is being requested. The final bits are about that fifo, which needs to wait until the end of the currently available logs because it's the last N and not the first N, and the follow flag, which becomes a breaking point to decide whether to keep going or abort. I think we can make the loop 100% agnostic and simplify it on the way by moving it to a go routine on its own function, and accepting a single cancellation channel as a parameter.

Something along those lines:

```
func ServeHTTP(...) {

    (...)

    logs := make(chan servicelog.Entry)
    go r.steamLogs(logs, requestDone)

    for {
        select {
        case log := <-logs:
            if log.Time.IsZero() {
                    if follow {
                        continue
                    }
                    break
            }
            if numLogs > 0 {
                 // ... add to fifo
            } else {
                // ... encode and write
            }
    }
    close(requestDone)
    // ... flush fifo
}

```

In other words, the entire logic inside the loop is now unaware of fifo or follow. It just sends logs out for as long as it can, until the stop channel is closed. These features are instead implemented at the call site. My theory at the moment is that the code inside the loop will then be simpler and more linear.

What do you think?

Reply:
That's a good suggestion. Strictly speaking we wouldn't need a goroutine, we could use some kind of `logStreamer` type with yet another `next()` method ... I tried this initially but IMO it made the code worse. So I switched to your goroutine/channel based approach, and I think it's pretty clean now. The `streamLogs` goroutine handles just the merging of log streams by timestamp, and the main loop handles the request-specific logic (`follow`, `numLog`).